staging/src/k8s.io/client-go/tools/portforward/tunneling_connection.go

```diff
@@ -0,0 +1,165 @@
+*/
+
+package portforward
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"net"
+	"time"
+
+	gwebsocket "github.com/gorilla/websocket"
+
+	"k8s.io/klog/v2"
+)
+
+const writeDeadline = 2 * time.Second
+
+var _ net.Conn = &TunnelingConnection{}
+
+// TunnelingConnection implements the "httpstream.Connection" interface, wrapping
+// a websocket connection that tunnels SPDY.
+type TunnelingConnection struct {
+	conn              *gwebsocket.Conn
+	closeChan         chan bool
+	maxBytesPerSec    int64
+	inProgressMessage io.Reader
+}
+
+// NewTunnelingConnection wraps the passed gorilla/websockets connection
+// with the TunnelingConnection struct (implementing net.Conn).
+func NewTunnelingConnection(conn *gwebsocket.Conn, maxBytesPerSec int64) *TunnelingConnection {
+	closeChan := make(chan bool)
+	tConn := &TunnelingConnection{
+		conn:           conn,
+		closeChan:      closeChan,
+		maxBytesPerSec: maxBytesPerSec,
+	}
+	// Close channel when detecting close connection.
+	closeHandler := conn.CloseHandler()
+	conn.SetCloseHandler(func(code int, text string) error {
+		klog.V(3).Infof("websocket conn close: %d--%s", code, text)
+		close(closeChan)
+		err := closeHandler(code, text)
+		return err
+	})
+	return tConn
+}
+
+func (c *TunnelingConnection) Read(p []byte) (int, error) {
+	klog.Infoln("tunneling connection read...")
+	defer klog.Infoln("tunneling connection read...complete")
+	for {
+		if c.inProgressMessage == nil {
+			klog.Infoln("tunneling connection read before NextReader()...")
+			messageType, nextReader, err := c.conn.NextReader()
+			if err != nil {
+				closeError := &gwebsocket.CloseError{}
+				if errors.As(err, &closeError) && closeError.Code == gwebsocket.CloseNormalClosure {
+					return 0, io.EOF
+				}
+				klog.Errorf("tunneling connection NextReader() error: %v", err)
+				return 0, err
+			}
+			if messageType != gwebsocket.BinaryMessage {
+				return 0, fmt.Errorf("invalid message type received")
+			}
+			c.inProgressMessage = nextReader
+		}
+
+		klog.Infoln("tunneling connection read in progress message...")
+		i, err := c.inProgressMessage.Read(p)
+		klog.Infof("tunneling connection read in progress message...%d bytes, (error: %v)", i, err)
+		switch {
+		case err == nil:
+			return i, nil
+		case err == io.EOF:
+			c.inProgressMessage = nil
+		case err != nil:
+			return i, err
+		}
+	}
+}
+
+func (c *TunnelingConnection) Write(p []byte) (int, error) {
+	klog.Infof("tunneling connection write: %d bytes", len(p))
+	defer klog.Infoln("tunneling connection write...complete")
+	if c.conn == nil {
+		return 0, fmt.Errorf("write on closed tunneling connection")
+	}
+	// err := c.SetWriteDeadline(time.Now().Add(writeDeadline))
+	// if err != nil {
+	// 	klog.V(4).Infof("Tunneling conn setting write deadline failed %v", err)
+	// 	return 0, err
+	// }
+	w, err := c.conn.NextWriter(gwebsocket.BinaryMessage)
+	if err != nil {
+		return 0, err
+	}
+	defer func() {
+		if w != nil {
+			w.Close() //nolint:errcheck
```

Comment:
can simplify this with a named return and conditionally setting the return error

func (c *TunnelingConnection) Write(p []byte) (n int, err error) {
...
defer func(){
  // close, which flushes the message
  closeErr := w.Close()
  if closeErr != nil && err == nil {
    // if closing/flushing errored and we weren't already returning an error, return the close error
    err = closeErr
  }
}()
}

Reply:
Fixed. Please have a look.